File: challenges/2023/01-trebuchet/main.py
import sys
from typing import Optional


def _debug(*args, **kwargs):
    kwargs["file"] = sys.stderr
    print(*args, **kwargs)


def parse(inp: str) -> list[str]:
    return inp.splitlines()


def one(inp: str) -> int:
    parsed = parse(inp)
    acc = 0
    for line in parsed:
        digits = list(filter(lambda x: x.isdigit(), line))
        assert len(digits) != 0, f"must have at least one digit per line: {line}"
        val = digits[0] + digits[-1]
        acc += int(val)
    return acc


TRANSFORMATIONS = {
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9",
}


def find_first(inp: str, opts: list[str]) -> Optional[str]:
    candidates = []
    for opt in opts:
        if (p := inp.find(opt)) != -1:
            candidates.append((p, opt))
    li = sorted(candidates, key=lambda x: x[0])
    if len(li) == 0:
        return None
    return li[0][1]


def two(inp: str) -> int:
    parsed = parse(inp)
    acc = 0

    search_values = list(map(str, range(0, 10))) + list(TRANSFORMATIONS.keys())
    reversed_search_values = ["".join(reversed(x)) for x in search_values]

    for line in parsed:
        first_digit = find_first(line, search_values)
        second_digit = find_first("".join(reversed(line)), reversed_search_values)
        assert (
            first_digit is not None and second_digit is not None
        ), f"must have at least one digit per line: {line}"

        # second digit will be the reversed form
        second_digit = "".join(reversed(second_digit))

        first_digit = TRANSFORMATIONS.get(first_digit, first_digit)
        second_digit = TRANSFORMATIONS.get(second_digit, second_digit)

        val = first_digit + second_digit
        acc += int(val)

    return acc


if __name__ == "__main__":
    if len(sys.argv) < 2 or sys.argv[1] not in ["1", "2"]:
        print("Missing day argument", file=sys.stderr)
        sys.exit(1)
    inp = sys.stdin.read().strip()
    if sys.argv[1] == "1":
        print(one(inp))
    else:
        print(two(inp))

File: challenges/2023/01-trebuchet/main.hs
import System.Environment
import System.Exit
import System.IO
import Data.Maybe
import Data.Char
import Control.Exception

type ChallengeReturn = Int

parse :: String -> [String]
parse = lines

one :: String -> ChallengeReturn
one inp = foldl
    (\acc line ->
        let digits = map (\x -> ord x - ord '0') (filter isDigit line) in
            assert ((length digits) /= 0) (
                acc + ((digits!!0) * 10) + (digits!!((length digits) - 1))
            )
    )
    0
    (parse inp)

two :: String -> ChallengeReturn
two inp = undefined

main :: IO ()
main = do args <- getArgs
          inp <- getContents
          _runFn (_selectFn args) inp

_selectFn :: [String] -> Maybe (String -> ChallengeReturn)
_selectFn ["1"] = Just one
_selectFn ["2"] = Just two
_selectFn _ = Nothing

_runFn :: Maybe (String -> ChallengeReturn) -> String -> IO ()
_runFn Nothing _ = _debug "Missing or invalid day argument" >> exitWith (ExitFailure 1)
_runFn (Just fn) inp = putStrLn (show (fn inp)) >> exitWith ExitSuccess

_debug :: String -> IO ()
_debug x = do hPutStrLn stderr x
